### PR TITLE
[WIP] Draft of OpenAPI 3 support

### PR DESCRIPTION
**Pull request by rinold**
## PR Purpose: Not for Merge
Doesn't provide complete OAS3 specification implementation (still most of it).
The actual purpose is to find out differences to discuss and decide on best approaches of implementing backward compatible solution.

#### Based on [Open API 3 Specification](https://swagger.io/specification)

I will describe changes of existing objects and/or provide links in comments to files containing them so that they could be discussed separately.

#### TODOs:
- Merge in backward compatible way
- "x-" extension of OAS3 objects (mostly any allowed)
- Any object decodable (for Example Objects and similar)
- Structure files: may be something like Common (for shared), OpenAPI2 (for api 2 specific objects/implementations), OpenAPI3 (for new api 3 objects/changed implementations)
- Add Tests and Examples (e.g. from openAPI3 git repo, the ones I've used for local testing purposes)

### CODE REVIEW

File: Sources/Callback.swift [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        -> Either<Callback, Structure<Callback>>
+    {
+        switch reference {
+        case .pointer(let pointer):
+            return .b(try self.resolver.resolve(swagger, pointer: pointer))
+        case .value(let builder):
+            return .a(try builder.build(swagger))
+        }
+    }
+}
```
Comment (rinold) [inline Sources/Callback.swift]: ### [Callback Object](https://swagger.io/specification/#callbackObject)

File: Sources/Components.swift [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                          responses: responses,
+                          parameters: parameters,
+                          examples: examples,
+                          requestBodies: requestBodies,
+                          headers: headers,
+                          securitySchemes: securitySchemes,
+                          links: links,
+                          callbacks: callbacks)
+    }
+}
```
Comment (rinold) [inline Sources/Components.swift]: ### Reusable [Components Object](https://swagger.io/specification/#componentsObject)
Used in **Swagger** field 'components'

File: Sources/Either.swift [hunk 3]
```diff
@@ -32,3 +32,17 @@ enum CodableEither<A: Codable, B: Codable>: Codable {
...
+    var structure: A {
+        switch self {
+        case .a(let vA):
+            return vA
+        case .b(let vB):
+            return vB.structure
+        }
+    }
+    
+}
```
Comment (rinold) [inline Sources/Either.swift:48]: Added this while testing references, just a convenience method to get A value of `Either<A, Structure<A>>`.

File: Sources/Encoding.swift [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    func build(_ swagger: SwaggerBuilder) throws -> Encoding {
+        let contentType = self.contentType ?? "" // TODO: Default values
+        let headers = try self.headers.mapValues { try HeaderBuilder.resolve(swagger, reference: $0) }
+        return Encoding(contentType: contentType,
+                        headers: headers,
+                        style: self.style,
+                        explode: self.explode ?? false, // TODO: default depends on style
+                        allowReserved: self.allowReserved ?? false)
+    }
+}
```
Comment (rinold) [inline Sources/Encoding.swift]: ### [Encoding Object](https://swagger.io/specification/#encodingObject)

File: Sources/Enums.swift [hunk 5]
```diff
@@ -1,10 +1,9 @@
 
 public enum ParameterLocation: String, Codable {
+    case path
     case query
     case header
-    case path
-    case formData
-    case body
+    case cookie
 }
```
Comment (rinold) [inline Sources/Enums.swift]: #### [Parameters Location](https://swagger.io/specification/#parameterObject) in version 3.0:

There are four possible parameter locations specified by the in field:

**path** - Used together with Path Templating, where the parameter value is actually part of the operation's URL. This does not include the host or base path of the API. For example, in /items/{itemId}, the path parameter is itemId.

**query** - Parameters that are appended to the URL. For example, in /items?id=###, the query parameter is id.

**header** - Custom headers that are expected as part of the request. Note that RFC7230 states header names are case insensitive.

**cookie** - Used to pass a specific cookie value to the API.

File: Sources/Enums.swift [hunk 6]
```diff
@@ -132,3 +132,13 @@ public enum DataType: String, Codable {
...
+
+public enum SerializationStyle: String, Codable {
+    case matrix
+    case label
+    case form
+    case simple
+    case spaceDelimited
+    case pipeDelimited
+    case deepObject
+}
```
Comment (rinold) [inline Sources/Enums.swift]: #### [Serialization Styles](https://swagger.io/specification/#style-values-52)

File: Sources/Example.swift [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        -> Either<Example, Structure<Example>>
+    {
+        switch reference {
+        case .pointer(let pointer):
+            return .b(try self.resolver.resolve(swagger, pointer: pointer))
+        case .value(let builder):
+            return .a(try builder.build(swagger))
+        }
+    }
+}
```
Comment (rinold) [inline Sources/Example.swift]: ### [Example Object](https://swagger.io/specification/#exampleObject)

File: Sources/Header.swift [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        -> Either<Header, Structure<Header>>
+    {
+        switch reference {
+        case .pointer(let pointer):
+            return .b(try self.resolver.resolve(swagger, pointer: pointer))
+        case .value(let builder):
+            return .a(try builder.build(swagger))
+        }
+    }
+}
```
Comment (rinold) [inline Sources/Header.swift]: ### [Header Object](https://swagger.io/specification/#headerObject)

This is quite the same as Parameter in OAS3, except that it excludes (with MUST) the 'name' and 'in' fields. Thus I've created the ParameterDefinition and it's builder which are reused here and in Parameter itself.

File: Sources/Link.swift [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        switch reference {
+        case .pointer(let pointer):
+            return .b(try self.resolver.resolve(swagger, pointer: pointer))
+        case .value(let builder):
+            return .a(try builder.build(swagger))
+        }
+    }
+}
+
```
Comment (rinold) [inline Sources/Link.swift]: ### [Link Object](https://swagger.io/specification/#linkObject)

File: Sources/MediaType.swift [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        let examples = try self.examples.mapValues {
+            try ExampleBuilder.resolve(swagger, reference: $0)
+        }
+        let encoding = try self.encoding.mapValues { try $0.build(swagger) }
+        return MediaType(schema: schema,
+                         example: self.example as Any?,
+                         examples: examples,
+                         encoding: encoding)
+    }
+}
```
Comment (rinold) [inline Sources/MediaType.swift]: ### [Media Type Object](https://swagger.io/specification/#mediaTypeObject)

File: Sources/OAuthFlows.swift [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+extension OAuth2FlowsBuilder: Builder {
+    typealias Building = OAuth2Flows
+    
+    func build(_ swagger: SwaggerBuilder) throws -> OAuth2Flows {
+        return OAuth2Flows(implicit: try implicit?.build(swagger),
+                           password: try password?.build(swagger),
+                           clientCredentials: try clientCredentials?.build(swagger),
+                           authorizationCode: try authorizationCode?.build(swagger))
+    }
+}
```
Comment (rinold) [inline Sources/OAuthFlows.swift]: ### [OAuth Flows Object](https://swagger.io/specification/#oauthFlowsObject)

File: Sources/Operation.swift [hunk 12]
```diff
@@ -1,101 +1,118 @@
...
+    public let callbacks: [String: Either<Callback, Structure<Callback>>]
+    
+    /// Declares this operation to be deprecated.
+    /// Consumers SHOULD refrain from usage of the declared operation.
     /// Default value is false.
-    public let deprecated: Bool
-
-    /// A unique string used to identify the operation
-    public let identifier: String?
+    public let deprecated: Bool?
```
Comment (rinold) [inline Sources/Operation.swift]: FIXME: Unnecessary change

File: Sources/Operation.swift [hunk 13]
```diff
@@ -116,21 +133,24 @@ extension OperationBuilder: Builder {
...
+                         externalDocumentation: externalDocumentation,
+                         identifier: self.identifier,
+                         parameters: parameters,
+                         requestBody: requestBody,
+                         responses: responses,
+                         callbacks: callbacks,
+                         deprecated: self.deprecated,
+                         security: self.security)
     }
 }
```
Comment (rinold) [inline Sources/Operation.swift]: ### [Operation Object](https://swagger.io/specification/#operationObject)

This object seems to differ in OAS2 and OAS3 - possibly we would need 2 different 'versioned' objects.

File: Sources/Parameter.swift [hunk 14]
```diff
@@ -1,61 +1,106 @@
...
+                         example: definition.example,
+                         examples: definition.examples,
+                         content: definition.content)
     }
 }
 
+
 extension ParameterBuilder {
     static func resolve(_ swagger: SwaggerBuilder, reference: Reference<ParameterBuilder>) throws
         -> Either<Parameter, Structure<Parameter>>
```
Comment (rinold) [inline Sources/Parameter.swift]: ### [Parameter Object](https://swagger.io/specification/#parameterObject)

As per new OAS3 Parameters Location values - body is now excluded (handled via separate Request Body Object).

File: Sources/Path.swift [hunk 15]
```diff
@@ -46,8 +62,13 @@ extension PathBuilder: Builder {
...
         let operations = try self.operations.mapValues { try $0.build(swagger) }
         let parameters = try self.parameters.map { try ParameterBuilder.resolve(swagger, reference: $0) }
-        return Path(operations: operations, parameters: parameters)
+        return Path(summary: self.summary,
+                    description: self.description,
+                    servers: servers,
+                    operations: operations,
+                    parameters: parameters)
     }
 }
```
Comment (rinold) [inline Sources/Path.swift]: ### [Path Item Object](https://swagger.io/specification/#pathItemObject)

Seems this object also should be version for v2 and v3 support.

File: Sources/ReferenceResolver.swift [hunk 16]
```diff
@@ -5,7 +5,7 @@ protocol ResolvableType: Builder {
 
 private let kSchemaBuilderReferenceResolver = ReferenceResolver<SchemaBuilder>()
 extension SchemaBuilder: ResolvableType {
-    static var path: String { return "definitions" }
+    static var path: String { return "schemas" }
```
Comment (rinold) [inline Sources/ReferenceResolver.swift]: The only different path in OAS2 vs OAS3 (except new/dropped) ones - we could add some versioning by changing it to func accepting swagger/version or possibly easier would be to create some duplicate of it like 'DefinitionBuilder'?

File: Sources/ReferenceResolver.swift [hunk 17]
```diff
@@ -53,22 +93,45 @@ class ReferenceResolver<T: ResolvableType> {
...
+            
+            // Push the resolving reference into the context to detect cyclic references.
+            context.resolvingReferences.insert(name)
+
+            return try resolve(swagger, pointer: pointer)
+        }
+        
+        guard case let Reference.value(builder) = builderRef else {
             throw ResolverError.unresolvedReference
         }
```
Comment (rinold) [inline Sources/ReferenceResolver.swift]: ### ReferenceResolver
- Path changed (e.g. '#/components/schemas/Tag' vs '#/shemas/Tag')
- Components object now contains Object or Reference, while previously reusable objects were not allowed to be references.

#### Merge proposal(s):
- Path could be handled by version check.
- Components now contains much more objects allowed for reuse and their type changed to Reference<T> instead of T. The referenced builders container also changed, thus possibly we could add protocol with method like `func referencedBuilder(for name: String) -> Any` or `Either<T, Reference<T>>` instead of `Any`, and support both versions in internal versioned if/else, still maintaining most of the code common.

File: Sources/RequestBody.swift [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+        -> Either<RequestBody, Structure<RequestBody>>
+    {
+        switch reference {
+        case .pointer(let pointer):
+            return .b(try self.resolver.resolve(swagger, pointer: pointer))
+        case .value(let builder):
+            return .a(try builder.build(swagger))
+        }
+    }
+}
```
Comment (rinold) [inline Sources/RequestBody.swift]: ### [Request Body Object](https://swagger.io/specification/#requestBodyObject)

File: Sources/Schema.swift [hunk 19]
```diff
@@ -46,3 +46,16 @@ extension SchemaBuilder: Builder {
...
+        -> Either<Schema, Structure<Schema>>
+    {
+        switch reference {
+        case .pointer(let pointer):
+            return .b(try self.resolver.resolve(swagger, pointer: pointer))
+        case .value(let builder):
+            return .a(try builder.build(swagger))
+        }
+    }
+}
```
Comment (rinold) [inline Sources/Schema.swift]: Shema now referencable.

File: Sources/SecuritySchema.swift [hunk 20]
```diff
@@ -68,6 +81,22 @@ extension SecuritySchemaBuilder: Builder {
...
+    {
+        switch reference {
+        case .pointer(let pointer):
+            return .b(try self.resolver.resolve(swagger, pointer: pointer))
+        case .value(let builder):
+            return .a(try builder.build(swagger))
+        }
+    }
+}
+
```
Comment (rinold) [inline Sources/SecuritySchema.swift]: ### [Security Schema Object](https://swagger.io/specification/#securitySchemeObject)

File: Sources/Swagger.swift [hunk 21]
```diff
@@ -9,50 +9,17 @@ public struct Swagger {
     /// Provides metadata about the API. The metadata can be used by the clients if needed.
```
Comment (rinold) [inline Sources/Swagger.swift]: ### Swagger
As per new specification of [OAS Object](https://swagger.io/specification/#oasObject):
- Version field 'swagger' replaced to 'openapi', with versioning (3.0.*)
- Reusable components moved to 'components' field ([Components Object](https://swagger.io/specification/#componentsObject))
- 'host' and 'basePath' replaced with 'servers' (array of [Server Object](https://swagger.io/specification/#serverObject))

#### Merge proposal(s):
Possibly, we can leave the swagger and openapi CodingKeys in top level Swagger structure, but depending on what key exists - create and build two separate implementation of e.g. OpenAPI2 and OpenAPI3 objects containing all other declaration.

File: Sources/Server.swift [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+extension ServerBuilder: Builder {
+    typealias Building = Server
+    
+    func build(_ swagger: SwaggerBuilder) throws -> Server {
+        let variables = self.variables?.mapValues({ try! $0.build(swagger) })
+        return Server(url: self.url,
+                      description: self.description,
+                      variables: variables)
+    }
+}
```
Comment (rinold) [inline Sources/Server.swift]: ### [Server Object](https://swagger.io/specification/#serverObject)

**Comment by AttilaTheFun** [comment]
Whoa @rinold this is awesome! I'll look it over this weekend

**Comment by AttilaTheFun** [comment]
I could make a openapi-3 branch that we could potentially merge this into as a staging area and merge that to master once it's fairly complete and backwards compatible

**Comment by rinold** [comment]
Regarding openapi-3 branch - I think it would be good idea. As for now, I'm thinking about the top-level API which could support both versions in a convenient usage manner. Will share my ideas a bit later, and looking forward to hearing yours!

**Comment by nxtstep** [comment]
@AttilaTheFun @rinold hey guys, this is really great work 🏆. Is there any chance this PR is getting merged soonish. It would be a shame if all this effort goest to waste. Or is there a better alternative to parse openAPI3 specs that I'm unaware of?

**Comment by AttilaTheFun** [comment]
Hey folks! Sorry I haven't looked at this in a long time. I would be happy to at least merge it into an openapi-3 branch and can do that now. I'd have to spend some time re-reviewing it to see how I feel about merging it into master and cutting a new version. The project I originally created this for has since moved to GraphQL. If you're interested in continuing to develop this within the main repo (and creating a PR I could review from the openapi-3 branch into master) I'd be happy to add you as collaborators.

cc @nxtstep @rinold

**Review by AttilaTheFun — APPROVED** [review]
I'm accepting this to merge into the openapi-3 branch but I'll have to spend some more time reviewing it before it merges to master.
